### DISCUSSION TITLE
[HELP] XtermJS changing cols and rows size but linux isn't

### DISCUSSION THREAD

**Discussion by Gigko** [sourcelair/xterm.js#5144] [Q&A] [answered]
I'm facing an issue with my Linux CentOS VM - CentOS Linux release 7.9.2009 (Core), I hope you can help me.
I'm successfully connecting to my VM from a web page via web socket.

Unfortunately, cols and rows size can be changed only manually in my case via command:
[root@server/]# tput cols
80
[root@server/]# stty cols 200 rows 40
[root@server/]# tput cols
200

Are there any options to change cols and rows size dynamically ?

**Comment by jerch** [comment]
This depends on your PTY integration, most PTY libs offer a way to change the PTY size via some API call (e.g. _node-pty_ has a _resize_ method).

**Reply by Gigko** [reply]
You meant ssh to web socket implementation? 
Can you suggest the most widely \ best option for this?

**Reply by jerch** [reply]
> You meant ssh to web socket implementation?

Nope, this has nothing to do with ssh. For a proper terminal experience you need to hook xterm.js to the master file descriptor of an actual PTY. This PTY you need to resize. Normally this is done via some PTY bridging lib, which also brings a resize call.

> Can you suggest the most widely \ best option for this?

Most widely used is prolly the C ioctl way. But seriously, this should already be encapsulated by your PTY lib. So which PTY lib do you use?

**Reply by Gigko** [reply]
> > You meant ssh to web socket implementation?
> 
> Nope, this has nothing to do with ssh. For a proper terminal experience you need to hook xterm.js to the master file descriptor of an actual PTY. This PTY you need to resize. Normally this is done via some PTY bridging lib, which also brings a resize call.
> 
> > Can you suggest the most widely \ best option for this?
> 
> Most widely used is prolly the C ioctl way. But seriously, this should already be encapsulated by your PTY lib. So which PTY lib do you use?

I'm not that deep developer, I meant to ask about proper tools I need to connect xtermjs to an a ssh session.

As far as I understood, I need to redo my way to share ssh session, can you suggest which tools to use to share ssh session via websocket?

**Reply by jerch** [reply]
In your first post you write

> I'm successfully connecting to my VM from a web page via web socket.

So the question is - how did you do that? Where do you get the data from to feed to xtermjs?

**Reply by Gigko** [reply]
> In your first post you write
> 
> > I'm successfully connecting to my VM from a web page via web socket.
> 
> So the question is - how did you do that?

I used websocket part from this project, https://github.com/zephinzer/cloudshell it worked well for test and debug, but when my project grew up I notice that user experience is not great because of the resizing issue. 

So now I understood that I need to find a proper way to share an ssh session via websocket.

**Reply by jerch** [reply]
You cannot directly share an ssh session (or more exactly an ssh IO stream) with a websocket. You'd need to do the following (schematically):
- spawn a PTY on some linux/posix machine
- run ssh there as first process (slave side of the PTY)
- connect xtermjs to master side of the PTY (e.g. with the help of websockets or ajax/fetch long polling)

The tricky part is the PTY spawning and getting its file descriptor setup right. Normally you would not do this yourself as it is quite involved with ugly kernel/C interfaces. Thus you should look for an appropriate lib doing this for you in your backend language. There are several libs for this in rust, Go, python or for nodejs available. If you use nodejs on server side, _node-pty_ is prolly the easiest for you to use.

**Reply by Gigko** [reply]
> You cannot directly share an ssh session (or more exactly an ssh IO stream) with a websocket. You'd need to do the following (schematically):
> 
> * spawn a PTY on some linux/posix machine
> * run ssh there as first process (slave side of the PTY)
> * connect xtermjs to master side of the PTY (e.g. with the help of websockets or ajax/fetch long polling)
> 
> The tricky part is the PTY spawning and getting its file descriptor setup right. Normally you would not do this yourself as it is quite involved with ugly kernel/C interfaces. Thus you should look for an appropriate lib doing this for you in your backend language. There are several libs for this in rust, Go, python or for nodejs available. If you use nodejs on server side, _node-pty_ is prolly the easiest for you to use.

Thank you!
I'll try to implement this tomorrow and I'll let you know the results. 

Anyways, to run my project right, I must dig and understand how this part of my system works.

**Reply by Gigko** [reply, marked as answer]
I found that https://github.com/cs01/pyxtermjs
Working better with terminal resize feature
